app/services/category_hashtag_data_source.rb

```diff
@@ -22,39 +20,27 @@ def self.category_to_hashtag_item(parent_category, category)
       # categories here.
       item.ref =
         if category.parent_category_id
-          !parent_category ? category.slug : "#{parent_category[:slug]}:#{category.slug}"
+          "#{category.parent_category.slug}:#{category.slug}"
         else
           category.slug
         end
     end
   end
 
   def self.lookup(guardian, slugs)
-    # We use Site here because it caches all the categories the
-    # user has access to.
-    guardian_categories = Site.new(guardian).categories
+    user_categories = Category.secured(guardian).includes(:parent_category)
     Category
-      .query_from_cached_categories(slugs, guardian_categories)
-      .map do |category|
-        parent_category =
-          guardian_categories.find { |cat| cat[:id] == category[:parent_category_id] }
-        category_to_hashtag_item(parent_category, category)
-      end
+      .query_from_cached_categories(slugs, user_categories)
+      .map { |category| category_to_hashtag_item(category) }
   end
 
   def self.search(guardian, term, limit)
-    guardian_categories = Site.new(guardian).categories
-
-    guardian_categories
-      .select do |category|
-        category[:name].downcase.include?(term) || category[:slug].downcase.include?(term)
-      end
+    Category
+      .secured(guardian)
+      .includes(:parent_category)
+      .where("LOWER(name) LIKE :term OR LOWER(slug) LIKE :term", term: "%#{term}%")
```

Comment:
Should we be matching mid-string? Or should we go for less noise and anchor to the start of the string (i.e. `"#{term}%"`, assuming users will start typing the category from the beginning?

Reply:
Anchoring at the start of the string is already planned for another commit :) Well, ranking by start of string at any rate!